### DISCUSSION TITLE
Agama flow and authorization code for the token enpoint

### DISCUSSION THREAD

**Discussion by toroarnaud** [JanssenProject/jans#6618] [Q&A] [answered]
Hi,

I am trying with an Agama flow to do an openid authentication as we do it with an interception script.
My Agama flow is working fine, I can create users and authenticate them. The flow is triggered with an openid authentication request of the form:
_https://<jans-server-name>/jans-auth/restv1/authorize?acr_values=agama&agama_flow=flow-qname&scope=...&response_type=...&redirect_uri=https...&client_id=..._
The problem is I don't see how to get the _**code**_ used to obtain an access token.
When using an interception script, after successful authentication of the user we are redirected to the _redirect_uri_ (passed as parameter to the _authorize_ endpoint) which includes the _authorization code_.

I am sure I am missing something in the logic when using the Agama flow and it would be nice if someone could tell me how to obtain this authentication code.

Thank you!

**Comment by ossdhaval** [comment]
Adding @jgomer2001 or @maduvena for comments.

**Comment by jgomer2001** [comment]
If your agama flow finished successfully, the browser will be taken to the redirect uri you passed when the flow was triggered. That one has to have to code

**Comment by toroarnaud** [comment]
Thank you so much for your reply!

I had a form displaying a "login success" message before the Finish statement and I believe it was the reason why the redirection was not occurring. I've removed this form from the flow and now the redirection takes place but ending up with this:

https://_FQDN_/admin?error_description=The+Authorization+Server+requires+End-User+authentication.+This+error+MAY+be+returned+when+the+prompt+parameter+in+the+Authorization+Request+is+set+to+none+to+request+that+the+Authorization+Server+should+not+display+any+user+interfaces+to+the+End-User%2C+but+the+Authorization+Request+cannot+be+completed+without+displaying+a+user+interface+for+user+authentication.+CorrelationId%3A+27cb0920-508e-432a-98f8-898cc404e483&hint=Create+authorization+request+to+start+new+authentication+session.&error=login_required

I've checked and there is no prompt parameter in the authorization request.
Before reaching the Finish state the authentication is done using:
**boolean isAuthenticated = authenticationService.authenticate(username, password);**
And isAuthenticated is true. 
Is there anything else to be done before being able to get successfully the code ?

**Comment by jgomer2001** [comment]
> Before reaching the Finish state the authentication is done using:
> boolean isAuthenticated = authenticationService.authenticate(username, password);

Agama does not work like jython scripts...

`Finish` directive is used to convey the identity of the person that will be authenticated (and terminate execution)

Check [this](https://docs.jans.io/head/admin/developer/agama/jans-agama-engine/#authentication-and-finish) for more info

**Comment by toroarnaud** [comment]
Thanks for the link!
In my finish endpoint I have this:
obj = { success: true, data: { userId: creds.username } }
It is still failing and in the jans-auth_script.log file I can see this:
(PythonService.java:243) - Agama. No user matches the required condition: uid=Arnaud  
However I do have a user in the jans_user table a user with "uid" = "Arnaud" so I don't know what's wrong...

And checking the jans-auth server logs:
Marking flow run as finished...
auth - 2023-11-24 17:18:22,167 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:56) - Retrieving current flow's status
auth - 2023-11-24 17:18:22,181 INFO  [qtp1492875057-22] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:230) - Removing flow run...
auth - 2023-11-24 17:18:22,192 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:223) - Getting user information from LDAP: attributeName = 'uid', attributeValue = 'Arnaud'
auth - 2023-11-24 17:18:22,212 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:235) - Found '0' entries

The fact that the user is not found explains the failure but the question is why the user is not found since it does exist in the DB. I've also tried with the admin user that is created when installing the server thinking it could be the way I've registered my User but the result is the same, it cannot be found.

By the way when I call the authenticate method in my Agama flow, the user is found...:
auth - 2023-11-24 17:28:20,823 DEBUG [qtp1492875057-1124] [io.jans.agama.engine.service.ActionService] (ActionService.java:91) - Class com.toro.passclub.BasicAuthentication loaded successfully
auth - 2023-11-24 17:28:20,823 DEBUG [qtp1492875057-1124] [io.jans.agama.engine.service.ActionService] (ActionService.java:139) - Method found: public static boolean com.toro.passclub.BasicAuthentication.authenticate(java.lang.String,java.lang.String)
auth - 2023-11-24 17:28:20,823 DEBUG [qtp1492875057-1124] [io.jans.agama.engine.service.ActionService] (ActionService.java:158) - Examining argument at index 0
auth - 2023-11-24 17:28:20,824 DEBUG [qtp1492875057-1124] [io.jans.agama.engine.service.ActionService] (ActionService.java:158) - Examining argument at index 1
auth - 2023-11-24 17:28:20,824 DEBUG [qtp1492875057-1124] [io.jans.agama.engine.service.ActionService] (ActionService.java:143) - Performing method call
auth - 2023-11-24 17:28:20,825 DEBUG [qtp1492875057-1124] [jans.as.common.service.common.UserService] (UserService.java:223) - Getting user information from LDAP: attributeName = 'uid', attributeValue = 'Arnaud'
auth - 2023-11-24 17:28:20,861 DEBUG [qtp1492875057-1124] [jans.as.common.service.common.UserService] (UserService.java:235) - Found '1' entries
auth - 2023-11-24 17:28:20,866 DEBUG [qtp1492875057-1124] [io.jans.as.server.service.AuthenticationService] (AuthenticationService.java:128) - Authenticating user with LDAP: username: 'Arnaud', credentials: '1194932016'
auth - 2023-11-24 17:28:20,866 DEBUG [qtp1492875057-1124] [jans.as.common.service.common.UserService] (UserService.java:78) - Getting user information from LDAP: userId = Arnaud
auth - 2023-11-24 17:28:20,877 DEBUG [qtp1492875057-1124] [jans.as.common.service.common.UserService] (UserService.java:93) - Found 1 entries for user id = Arnaud
auth - 2023-11-24 17:28:20,883 DEBUG [qtp1492875057-1124] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:198) - Retrieving flow run 1d9ecaf9-7783-4340-9442-71f3f3b5e5c4
auth - 2023-11-24 17:28:20,894 INFO  [qtp1492875057-1124] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:219) - Marking flow run as finished...
auth - 2023-11-24 17:28:21,968 DEBUG [qtp1492875057-1127] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:56) - Retrieving current flow's status
auth - 2023-11-24 17:28:21,982 INFO  [qtp1492875057-1127] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:230) - Removing flow run...
auth - 2023-11-24 17:28:21,992 DEBUG [qtp1492875057-1127] [jans.as.common.service.common.UserService] (UserService.java:223) - Getting user information from LDAP: attributeName = 'uid', attributeValue = 'admin'
auth - 2023-11-24 17:28:22,012 DEBUG [qtp1492875057-1127] [jans.as.common.service.common.UserService] (UserService.java:235) - Found '0' entries
auth - 2023-11-24 17:28:22,014 DEBUG [qtp1492875057-1127] [io.jans.as.server.auth.Authenticator] (Authenticator.java:355) - Authentication result for user 'null'. auth_step: '1', result: 'false', credentials: '1194932016'
auth - 2023-11-24 17:28:22,014 DEBUG [qtp1492875057-1127] [io.jans.as.server.auth.Authenticator] (Authenticator.java:369) - Get next step from script: '-1'
auth - 2023-11-24 17:28:22,020 INFO  [qtp1492875057-1127] [io.jans.as.server.auth.Authenticator] (Authenticator.java:247) - Authentication failed for 'null'
auth - 2023-11-24 17:28:22,020 DEBUG [qtp1492875057-1127] [io.jans.as.model.error.ErrorResponseFactory] (ErrorResponseFactory.java:82) - Looking for the error with id: login_required

So basically when authenticating using Java code inside the Agama flow there is no issue, the user is found and authenticated but when the flow is marked as finished the user cannot be found for the same uid.

**Comment by toroarnaud** [comment]
Sorry in the above logs I've mixed both users Arnaud and admin between the authentication and flow finish for testing. Here are the proper logs with no mix up (in bold for both lookup in DB):
 
uth - 2023-11-24 17:56:29,767 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.FlowService] (FlowService.java:153) - Resuming flow
auth - 2023-11-24 17:56:29,871 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:91) - Class com.toro.passclub.BasicAuthentication loaded successfully
auth - 2023-11-24 17:56:29,872 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:139) - Method found: public static void com.toro.passclub.BasicAuthentication.testNoArg()
auth - 2023-11-24 17:56:29,872 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:143) - Performing method call
auth - 2023-11-24 17:56:29,882 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:91) - Class com.toro.passclub.BasicAuthentication loaded successfully
auth - 2023-11-24 17:56:29,882 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:139) - Method found: public static boolean com.toro.passclub.BasicAuthentication.authenticate(java.lang.String,java.lang.String)
auth - 2023-11-24 17:56:29,882 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:158) - Examining argument at index 0
auth - 2023-11-24 17:56:29,882 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:158) - Examining argument at index 1
auth - 2023-11-24 17:56:29,882 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.ActionService] (ActionService.java:143) - Performing method call
auth - 2023-11-24 17:56:29,883 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:223) - **Getting user information from LDAP: attributeName = 'uid', attributeValue = 'Arnaud'**
auth - 2023-11-24 17:56:29,904 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:235) - **Found '1' entries**
auth - 2023-11-24 17:56:29,909 DEBUG [qtp1492875057-22] [io.jans.as.server.service.AuthenticationService] (AuthenticationService.java:128) - Authenticating user with LDAP: username: 'Arnaud', credentials: '1194932016'
auth - 2023-11-24 17:56:29,909 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:78) - Getting user information from LDAP: userId = Arnaud
auth - 2023-11-24 17:56:29,920 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:93) - Found 1 entries for user id = Arnaud
auth - 2023-11-24 17:56:29,927 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:198) - Retrieving flow run f39a0103-3434-48b9-b8d2-81e8c93796d0
auth - 2023-11-24 17:56:29,937 INFO  [qtp1492875057-22] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:219) - Marking flow run as finished...
auth - 2023-11-24 17:56:31,065 DEBUG [qtp1492875057-1127] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:56) - Retrieving current flow's status
auth - 2023-11-24 17:56:31,086 INFO  [qtp1492875057-1127] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:230) - Removing flow run...
auth - 2023-11-24 17:56:31,165 DEBUG [qtp1492875057-1127] [jans.as.common.service.common.UserService] (UserService.java:223) - **Getting user information from LDAP: attributeName = 'uid', attributeValue = 'Arnaud'**
auth - 2023-11-24 17:56:31,185 DEBUG [qtp1492875057-1127] [jans.as.common.service.common.UserService] (UserService.java:235) - **Found '0' entries**
auth - 2023-11-24 17:56:31,188 DEBUG [qtp1492875057-1127] [io.jans.as.server.auth.Authenticator] (Authenticator.java:355) - Authentication result for user 'null'. auth_step: '1', result: 'false', credentials: '1194932016'
auth - 2023-11-24 17:56:31,189 DEBUG [qtp1492875057-1127] [io.jans.as.server.auth.Authenticator] (Authenticator.java:369) - Get next step from script: '-1'
auth - 2023-11-24 17:56:31,197 INFO  [qtp1492875057-1127] [io.jans.as.server.auth.Authenticator] (Authenticator.java:247) - Authentication failed for 'null'
auth - 2023-11-24 17:56:31,197 DEBUG [qtp1492875057-1127] [io.jans.as.model.error.ErrorResponseFactory] (ErrorResponseFactory.java:82) - Looking for the error with id: login_required
auth - 2023-11-24 17:56:31,197 DEBUG [qtp1492875057-1127] [io.jans.as.model.error.ErrorResponseFactory] (ErrorResponseFactory.java:87) - Found error, id: login_required

**Comment by nynymike** [comment]
That's very weird how in it can't find the user in the second call. 

The sequence is still a little unclear. When the user gets authenticated at the external OP, does this user already exist in the Auth Server database? Are you still not able to get the code?

**Comment by toroarnaud** [comment]
Hi,

I still cannot get the code and left it on hold because I have no clue on how to solve it.
There is no external OP and the user exists in the Auth Server database before the request is triggered.
I've attached an image if the Agama flow which is very simple: 
![agama_test_flow](https://github.com/JanssenProject/jans/assets/19341772/2dcdbb42-bbb3-4583-a14e-aff265e555b6)

- one form to collect the  username and password of the user to be authenticated
- A call to a Java class to authenticate the user with the following code:
  boolean isAuthenticated = authenticationService.authenticate(username, password)
  And I can see in the auth-script.log that isAuthenticated is true.

**Comment by duttarnab** [comment]
Hi Arnaud

Could you please share the `.gama` file or your agama project GH link so that we can take a look?

**Comment by toroarnaud** [comment]
Hi Arnab

Here it is:
[espanyol-login.zip](https://github.com/JanssenProject/jans/files/13707920/espanyol-login.zip)

THX!

**Comment by duttarnab** [comment]
Hi Arnaud

I have tested your `.gama` file. The flow is working for me. I created a new user in Janssen and tested with it. 
- It is able to generate the `code`
- It is able to find the user from persistence.

Attaching a video link here for your reference.
https://drive.google.com/file/d/1QmT5yia-VG_ObppYj0YKeDVPji5YaE-I/view?usp=sharing

Let me know if there is any concern.

Regards,

**Reply by toroarnaud** [reply]
Hi Arnab,

Thank you very much for trying!
Unfortunately, I still don't understand why it is not working in my case... The difference I see is the user used to test and the openid URL (I am not using neither a _code_challenge_ nor a _nonce_ but this should not impact I believe). In my case the user is the Admin user created by default when installing the jans_auth server:
{
  "displayName": "Default Admin User",
  "dn": "inum=296bb2f3-c001-448e-bc8e-441a3dffb00b,ou=people,o=jans",
  "emailVerified": true,
  "givenName": "Admin",
  "inum": "296bb2f3-c001-448e-bc8e-441a3dffb00b",
  "jansAdminUIRole": [
    "api-admin"
  ],
  "jansStatus": "active",
  "mail": "admin@kube-dev.passclub.es",
  "memberOf": [
    "inum=60B7,ou=groups,o=jans"
  ],
  "middleName": "Admin",
  "nickname": "Admin",
  "objectClass": "jansPerson",
  "sn": "User",
  "uid": "admin",
  "userPassword": "{ssha}BYdRN3kDGMNnmF+YYU+whJqcXUABFaLBJq2Zag=="
} 
and the url I'm using is:
https://kube-dev.passclub.es/jans-auth/authorize.htm?response_type=code&redirect_uri=https%3A%2F%2Fkube-dev.passclub.es%2Fadmin&client_id=1800.cc265ff0-cbcf-485f-97b8-cbba70ac5ac9&scope=openid&acr_values=agama&agama_flow=com.toro.passclub.espanyol

**And what is really weird** is that the user is found once (triggered by the script) and then the search fails while it is the same UserService class and method used for the search:

auth - 2023-12-19 10:48:20,833 DEBUG [qtp1492875057-21] [io.jans.agama.engine.service.ActionService] (ActionService.java:91) - Class com.toro.passclub.BasicAuthentication loaded successfully
auth - 2023-12-19 10:48:20,833 DEBUG [qtp1492875057-21] [io.jans.agama.engine.service.ActionService] (ActionService.java:139) - Method found: public static boolean com.toro.passclub.BasicAuthentication.authenticate(java.lang.String,java.lang.String)
auth - 2023-12-19 10:48:20,833 DEBUG [qtp1492875057-21] [io.jans.agama.engine.service.ActionService] (ActionService.java:158) - Examining argument at index 0
auth - 2023-12-19 10:48:20,834 DEBUG [qtp1492875057-21] [io.jans.agama.engine.service.ActionService] (ActionService.java:158) - Examining argument at index 1
auth - 2023-12-19 10:48:20,834 DEBUG [qtp1492875057-21] [io.jans.agama.engine.service.ActionService] (ActionService.java:143) - Performing method call
**auth - 2023-12-19 10:48:20,834 DEBUG [qtp1492875057-21] [jans.as.common.service.common.UserService] (UserService.java:223) - Getting user information from LDAP: attributeName = 'uid', attributeValue = 'admin'
auth - 2023-12-19 10:48:20,852 DEBUG [qtp1492875057-21] [jans.as.common.service.common.UserService] (UserService.java:235) - Found '1' entries**
auth - 2023-12-19 10:48:20,853 DEBUG [qtp1492875057-21] [io.jans.as.server.service.AuthenticationService] (AuthenticationService.java:128) - Authenticating user with LDAP: username: 'admin', credentials: '1795924650'
auth - 2023-12-19 10:48:20,853 DEBUG [qtp1492875057-21] [jans.as.common.service.common.UserService] (UserService.java:78) - Getting user information from LDAP: userId = admin
auth - 2023-12-19 10:48:20,862 DEBUG [qtp1492875057-21] [jans.as.common.service.common.UserService] (UserService.java:93) - Found 1 entries for user id = admin
auth - 2023-12-19 10:48:20,868 DEBUG [qtp1492875057-21] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:198) - Retrieving flow run 5315f531-312f-457b-9a72-c4a084dc5bce
auth - 2023-12-19 10:48:20,878 INFO  [qtp1492875057-21] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:219) - Marking flow run as finished...
auth - 2023-12-19 10:48:22,013 DEBUG [qtp1492875057-22] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:56) - Retrieving current flow's status
auth - 2023-12-19 10:48:22,022 INFO  [qtp1492875057-22] [io.jans.agama.engine.service.AgamaPersistenceService] (AgamaPersistenceService.java:230) - Removing flow run...
**auth - 2023-12-19 10:48:22,032 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:223) - Getting user information from LDAP: attributeName = 'uid', attributeValue = 'admin'
auth - 2023-12-19 10:48:22,052 DEBUG [qtp1492875057-22] [jans.as.common.service.common.UserService] (UserService.java:235) - Found '0' entries**
auth - 2023-12-19 10:48:22,055 DEBUG [qtp1492875057-22] [io.jans.as.server.auth.Authenticator] (Authenticator.java:355) - Authentication result for user 'null'. auth_step: '1', result: 'false', credentials: '1795924650'

Regards,
Arnaud

**Comment by duttarnab** [comment]
> And what is really weird is that the user is found once (**triggered by the script**) and **then the search fails** while it is the same UserService class and method used for the search:

- **triggered by the script** --- Do you mean agama script? 
-  **then the search fails** --- Have you written any other code to search the user again after authentication? I am asking because `UserService.java` is trying to get user information twice. In the second attempt, it failed. When I tested the `.gama` file user information was fetched only once.

I am not able to access your authorization url so not able to understand how it is working in your case. It would be helpful if you could attach an image of the page shown after authentication.

**Reply by toroarnaud** [reply]
**_triggered by the script --- Do you mean agama script?_**
Yes. The agama script (BasicAuthentication) triggers a seach which is successful.

**_then the search fails --- Have you written any other code to search the user again after authentication? I am asking because UserService.java is trying to get user information twice. In the second attempt, it failed. When I tested the .gama file user information was fetched only once._**
Ah ? No. I haven't written any other code. I am triggering the agama script which is the one you are testing too. So the question basically is "Where this second attempt comes from ?". What could trigger this second search ?

_**It would be helpful if you could attach an image of the page shown after authentication.**_
I reran the test with the same URL redirection that you are using (that is https://admin-test-ui.gluu.org) and I get the same OK page but with this redirection:
https://admin-ui-test.gluu.org/?error_description=The+Authorization+Server+requires+End-User+authentication.+This+error+MAY+be+returned+when+the+prompt+parameter+in+the+Authorization+Request+is+set+to+none+to+request+that+the+Authorization+Server+should+not+display+any+user+interfaces+to+the+End-User%2C+but+the+Authorization+Request+cannot+be+completed+without+displaying+a+user+interface+for+user+authentication.+CorrelationId%3A+425763f4-1295-4c33-84c2-5d873cfbbf93&hint=Create+authorization+request+to+start+new+authentication+session.&error=login_required

Instead of the redirection with the code.

**Reply by toroarnaud** [reply]
Also I don't know if this helps but if I enter a wrong password the low is exactly the same.

From the jans-auth-script.log file:

_Password is correct_:
2023-12-19 14:10:14,467 INFO  [qtp1492875057-16] 173a11fd-39c1-4238-b79c-b725b394eea4 [org.codehaus.groovy.vmplugin.v8.IndyInterface] (IndyInterface.java:318) - Passclub BasicAuthentication. **is authenticated: true**
2023-12-19 14:10:15,673 INFO  [qtp1492875057-15] eb72809f-f46e-4fe8-9def-fbd239c5714b [io.jans.service.PythonService$PythonLoggerOutputStream] (PythonService.java:243) - Agama. Authenticate for step 1
2023-12-19 14:10:15,692 INFO  [qtp1492875057-15] eb72809f-f46e-4fe8-9def-fbd239c5714b [io.jans.service.PythonService$PythonLoggerOutputStream] (PythonService.java:243) - Agama. Flow finished successfully
2023-12-19 14:10:15,712 INFO  [qtp1492875057-15] eb72809f-f46e-4fe8-9def-fbd239c5714b [io.jans.service.PythonService$PythonLoggerOutputStream] (PythonService.java:243) - **Agama. No user matches the required condition: uid=admin**

_Password is wrong_
2023-12-19 14:10:50,233 INFO  [qtp1492875057-16] 1b660692-6f84-4c91-92f9-2d75d8a0d642 [org.codehaus.groovy.vmplugin.v8.IndyInterface] (IndyInterface.java:318) - Passclub BasicAuthentication. **is authenticated: false**
2023-12-19 14:10:51,395 INFO  [qtp1492875057-20] a419224e-8f74-4eff-b19b-3873b5aa23f5 [io.jans.service.PythonService$PythonLoggerOutputStream] (PythonService.java:243) - Agama. Authenticate for step 1
2023-12-19 14:10:51,413 INFO  [qtp1492875057-20] a419224e-8f74-4eff-b19b-3873b5aa23f5 [io.jans.service.PythonService$PythonLoggerOutputStream] (PythonService.java:243) - Agama. Flow finished successfully
2023-12-19 14:10:51,433 INFO  [qtp1492875057-20] a419224e-8f74-4eff-b19b-3873b5aa23f5 [io.jans.service.PythonService$PythonLoggerOutputStream] (PythonService.java:243) - **Agama. No user matches the required condition: uid=admin**

**Comment by duttarnab** [comment]
I am not able to reproduce this error  :-|. I am getting authorization `code` after redirection. (as demonstrated in the video).

https://admin-ui-test.gluu.org/?code=fcb41ab0-2a5b-485b-a6e1-98ea0715affc&acr_values=agama&scope=openid&session_state=baab317bd6fdc324595a4c8edc7587f1529f4090e9bd5e3241f1795fd53ae91a.4aaa9474-97ef-48dd-ab59-1d4001ca1e67

![image](https://github.com/JanssenProject/jans/assets/32794267/229c1abf-a61f-4b2b-aa0d-1bb1a45492e9)

**Comment by nynymike** [comment]
Are we sure this client is correct? 

Can you export the client JSON and paste it here?  From the TUI you can do this by pressing `d` when you a viewing the list of clients, like this: 
![image](https://github.com/JanssenProject/jans/assets/3717101/5c47b2dc-9cd0-458e-b47f-2fdd7483b9e8)
Then you can export this client JSON to a file and paste here...

**Reply by toroarnaud** [reply]
Please find it in attachment
[client.json](https://github.com/JanssenProject/jans/files/13718830/client.json)

**Comment by duttarnab** [comment]
@toroarnaud , how old is your jans insatallation? There is a release happening every week so I would recommend testing with the latest installation.

You can check on my temporary installation and let me know if it works for you. Here is the authorization url. 

```
https://jv18creator-golden-crane.gluu.info/jans-auth/restv1/authorize?scope=openid&response_type=code&redirect_uri=https%3A%2F%2Fjv18creator-golden-crane.gluu.info&client_id=e26c0b1f-e6cb-40cc-b110-646d2bba9d12&code_challenge_method=S256&code_challenge=IaB1ZTZRcUjQOB7mWjv4DOVLdnFE8yH8pc7Kv4AVbeQ&nonce=b83529e2-a87b-4040-a736-a5566473943b&acr_values=agama&agama_flow=com.toro.passclub.espanyol
```
username: admin
password: Gluu1234.

**Reply by ossdhaval** [reply]
@toroarnaud As suggested by @duttarnab, have you tried the same scenario on recent builds? Do you still see the problem? Please post an update here if you are still pursuing this. Else I would like to close this discussion in a couple of days more.

**Comment by ALBDN** [comment]
Hi and first sorry for the late reply!
The server was installed using the helm chart with the default version to the tag that is 1.0.22_dev

**Reply by moabu** [reply, marked as answer]
Thats why. Please install using the latest.
